File: rtsp/src/main/java/com/pedro/rtsp/rtp/sockets/RtpSocketTcp.java
package com.pedro.rtsp.rtp.sockets;

import android.util.Log;
import com.pedro.rtsp.rtsp.RtpFrame;
import com.pedro.rtsp.utils.RtpConstants;
import java.io.IOException;
import java.io.OutputStream;

/**
 * Created by pedro on 7/11/18.
 */

public class RtpSocketTcp extends BaseRtpSocket {

  private OutputStream outputStream;
  private byte tcpHeader[];

  public RtpSocketTcp() {
    tcpHeader = new byte[] { '$', 0, 0, 0 };
  }

  @Override
  public void setDataStream(OutputStream outputStream, String host) {
    this.outputStream = outputStream;
  }

  @Override
  public void sendFrame(RtpFrame rtpFrame, boolean isEnableLogs) throws IOException {
    sendFrameTCP(rtpFrame, isEnableLogs);
  }

  @Override
  public void close() {

  }

  private void sendFrameTCP(RtpFrame rtpFrame, boolean isEnableLogs) throws IOException {
    synchronized (RtpConstants.lock) {
      int len = rtpFrame.getLength();
      tcpHeader[1] = rtpFrame.getChannelIdentifier();
      tcpHeader[2] = (byte) (len >> 8);
      tcpHeader[3] = (byte) (len & 0xFF);
      outputStream.write(tcpHeader);
      outputStream.write(rtpFrame.getBuffer(), 0, len);
      outputStream.flush();
      if (isEnableLogs) {
        Log.i(TAG, "wrote packet: "
            + (rtpFrame.getChannelIdentifier() == (byte) 2 ? "Video" : "Audio")
            + ", size: "
            + rtpFrame.getLength());
      }
    }
  }
}

File: rtsp/src/main/java/com/pedro/rtsp/rtp/sockets/RtpSocketTcp.kt
/*
 * Copyright (C) 2024 pedroSG94.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pedro.rtsp.rtp.sockets

import com.pedro.common.socket.TcpStreamSocket
import com.pedro.rtsp.rtsp.RtpFrame
import java.io.IOException

/**
 * Created by pedro on 7/11/18.
 */
class RtpSocketTcp : BaseRtpSocket() {

  private var socket: TcpStreamSocket? = null
  private val tcpHeader: ByteArray = byteArrayOf('$'.code.toByte(), 0, 0, 0)

  @Throws(IOException::class)
  override suspend fun setSocket(socket: TcpStreamSocket) {
    this.socket = socket
  }

  @Throws(IOException::class)
  override suspend fun sendFrame(rtpFrame: RtpFrame) {
    sendFrameTCP(rtpFrame)
  }

  override suspend fun flush() {
    socket?.flush()
  }

  override suspend fun close() {}

  @Throws(IOException::class)
  private suspend fun sendFrameTCP(rtpFrame: RtpFrame) {
    val len = rtpFrame.length
    tcpHeader[1] = (2 * rtpFrame.channelIdentifier).toByte()
    tcpHeader[2] = (len shr 8).toByte()
    tcpHeader[3] = (len and 0xFF).toByte()
    socket?.write(tcpHeader)
    socket?.write(rtpFrame.buffer, 0, len)
  }
}